### PR TITLE
regex: support partial matching in RE2

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: Xie Zhihao <zhihao.xie@intel.com>

Commit Message: regex: support partial matching in RE2
Additional Description:

Earlier this year, a user asks to support partial match in RE2 in #19948. The user finds difficulty to migrate API gateway from Nginx to Envoy since they have many regex patterns in data processing, while the Nginx's regex engine (and most other proxies and WAFs using PCRE-like regex engine) matches any substring in the input against the regex pattern, which is the same behavior of partial matching in RE2.

Previously, we can modify the regex pattern `/A/` to `/.*A.*/` as a workaround, but since we have implemented regex engine interface, we can add an option to enable partial matching globally. It will be more friendly for our users.

Risk Level: Low
Testing: Integration
Docs Changes: N/A
Release Notes: Added
Platform Specific Features: N/A

### CODE REVIEW

File: api/envoy/extensions/regex_engines/v3/google_re2.proto [hunk 1]
```diff
@@ -21,4 +21,8 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // which records the program size, and the counter ``re2.exceeded_warn_level``, which is incremented
 // each time the program size exceeds the warn level threshold.
 message GoogleRE2 {
+  // Allow to return matched if a substring of the input matches the regex pattern. By default, the
+  // engine will return matched only if the input matches the regex pattern fully, from the
+  // beginning to the end of the input.
+  bool partial_match = 1;
```
Comment (mattklein123) [inline api/envoy/extensions/regex_engines/v3/google_re2.proto:27]: Please add a release note.

File: source/common/common/regex.h [hunk 2]
```diff
@@ -40,13 +40,30 @@ class CompiledGoogleReMatcher : public CompiledMatcher {
...
+
+  // CompiledGoogleReMatcher
+  bool match(absl::string_view value) const override {
+    return re2::RE2::PartialMatch(re2::StringPiece(value.data(), value.size()), regex_);
+  }
+};
+
 class GoogleReEngine : public Engine {
 public:
+  GoogleReEngine();
```
Comment (mattklein123) [inline source/common/common/regex.h:60]: Why is the default constructor needed?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this. This seems OK, though at a high level I'm wondering if partial match regex should actually be a property of the higher layer matcher config as I'm guessing all engines would support this? This would be more user friendly if that's ultimately why we are doing this. cc @jmarantz for thoughts.

/wait

File: source/common/common/regex.h (hunk 2 above)
Comment (zhxie) [reply source/common/common/regex.h:60]: Oh, it is not needed. I just attempted to pass the CI and see if there if further issues in the patch. I will fix that.

File: api/envoy/extensions/regex_engines/v3/google_re2.proto (hunk 1 above)
Comment (zhxie) [reply api/envoy/extensions/regex_engines/v3/google_re2.proto:27]: Added, thanks.

**Comment by markdroth** [comment]
> though at a high level I'm wondering if partial match regex should actually be a property of the higher layer matcher config as I'm guessing all engines would support this? This would be more user friendly if that's ultimately why we are doing this.

FWIW, I agree with this.  I think this option would make a lot more sense as a field in the `RegexMatcher` proto.

**Comment by mattklein123** [comment]
> FWIW, I agree with this. I think this option would make a lot more sense as a field in the RegexMatcher proto.

OK let's go with this. @zhxie can you rework this PR? Thank you.

/wait

**Comment by tony612** [comment]
@zhxie Any change to merge this PR?

**Pull request closed without merging**